cmd/kubeadm/app/cmd/upgrade/node.go

```diff
@@ -19,221 +19,253 @@ package upgrade
+}
 
-	client, err := getClient(flags.kubeConfigPath, flags.dryRun)
-	if err != nil {
-		return errors.Wrapf(err, "couldn't create a Kubernetes client from file %q", flags.kubeConfigPath)
-	}
+// Cfg returns initConfiguration.
+func (d *nodeData) Cfg() *kubeadmapi.InitConfiguration {
+	return d.cfg
+}
 
-	waiter := apiclient.NewKubeWaiter(client, upgrade.UpgradeManifestTimeout, os.Stdout)
+// IsControlPlaneNode returns the isControlPlaneNode flag.
+func (d *nodeData) IsControlPlaneNode() bool {
+	return d.isControlPlaneNode
+}
 
-	// Fetches the cluster configuration
-	cfg, err := configutil.FetchInitConfigurationFromCluster(client, os.Stdout, "upgrade", false)
-	if err != nil {
-		return errors.Wrap(err, "unable to fetch the kubeadm-config ConfigMap")
-	}
+// Client returns a Kubernetes client to be used by kubeadm.
+func (d *nodeData) Client() clientset.Interface {
+	return d.client
+}
+
+// NewCmdUpgradeNodeConfig returns the cobra.Command for downloading the new/upgrading the kubelet configuration from the kubelet-config-1.X
+// ConfigMap in the cluster
+func NewCmdUpgradeNodeConfig() *cobra.Command {
+	nodeOptions := newNodeOptions()
+	nodeRunner := workflow.NewRunner()
+
+	cmd := &cobra.Command{
+		Use:        "config",
+		Short:      "Download the kubelet configuration from the cluster ConfigMap kubelet-config-1.X, where X is the minor version of the kubelet",
+		Deprecated: "use \"kubeadm upgrade node\" instead",
+		Run: func(cmd *cobra.Command, args []string) {
+			// This is required for preserving the old behavior of `kubeadm upgrade node config`.
+			// The new implementation exposed as a phase under `kubeadm upgrade node` infers the target
+			// kubelet config version from the kubeadm-config ConfigMap
+			if len(nodeOptions.kubeletVersion) == 0 {
+				kubeadmutil.CheckErr(errors.New("the --kubelet-version argument is required"))
+			}
 
-	// Upgrade the control plane and etcd if installed on this node
-	fmt.Printf("[upgrade] Upgrading your Static Pod-hosted control plane instance to version %q...\n", cfg.KubernetesVersion)
-	if flags.dryRun {
-		return DryRunStaticPodUpgrade(cfg)
+			err := nodeRunner.Run(args)
+			kubeadmutil.CheckErr(err)
+		},
 	}
 
-	if err := PerformStaticPodUpgrade(client, waiter, cfg, flags.etcdUpgrade, flags.renewCerts); err != nil {
-		return errors.Wrap(err, "couldn't complete the static pod upgrade")
+	// adds flags to the node command
+	addUpgradeNodeFlags(cmd.Flags(), nodeOptions)
+
+	// initialize the workflow runner with the list of phases
+	nodeRunner.AppendPhase(phases.NewKubeletConfigPhase())
+
+	// sets the data builder function, that will be used by the runner
+	// both when running the entire workflow or single phases
+	nodeRunner.SetDataInitializer(func(cmd *cobra.Command, args []string) (workflow.RunData, error) {
+		return newNodeData(cmd, args, nodeOptions)
+	})
+
+	return cmd
+}
+
+// NewCmdUpgradeControlPlane returns the cobra.Command for upgrading the controlplane instance on this node
```

Comment:
add TODO to remove when 1.18 is released?

Reply:
Now it is plugged in the command line
Even if it is alpha, the approach I'm proposing is to keep this around for 1 cycle, so we are not breaking users

Reply:
+1 for keeping even if alpha. 
we are GA and alpha will still break users.

Reply:
Probably it is necessary to make more clear what is GA and what not